src/scope.rs

```diff
@@ -156,6 +157,42 @@ impl BasePattern {
             _ => None,
         }
     }
+
+    /// Check if the `identifier` is contained inside the pattern.
+    pub fn contains(&self, identifier: &Identifier) -> bool {
+        self.pre_order_iter().any(|sub_pattern| {
+            sub_pattern
+                .as_identifier()
+                .map(|sub_id| sub_id == identifier)
+                .unwrap_or(false)
+        })
+    }
+
+    /// Compute a Simplicity expression that returns the value of the given `identifier`.
+    /// The expression takes as input a value that matches the `self` pattern.
+    ///
+    /// The expression is a sequence of `take` and `drop` followed by `iden`.
+    fn get(mut self: &Self, identifier: &Identifier) -> Option<SelectorBuilder<ProgNode>> {
+        let mut selector = SelectorBuilder::default();
+        loop {
+            // Termination: self becomes strictly smaller in each iteration
+            match self {
+                BasePattern::Identifier(self_id) if self_id == identifier => return Some(selector),
+                BasePattern::Identifier(_) | BasePattern::Ignore => return None,
+                BasePattern::Product(self_left, self_right) => {
+                    if self_left.contains(identifier) {
```

Comment:
In fb212840cc393d16ed2961721eb93650bec3c8d8:

I agree that the old `DirectedTree` code was overly general and hard to read. But this code has worse asymptotics because every call to `contains` iterates through the entire (sub)tree, so the algorithm winds up being O(n^2). It would be better if `contains` would record the path that it takes so that we don't need to call it more than once.

I would suggest using `verbose_pre_order_iter` in `contains` , and returning a vector of left/right decisions which you can pop every time you hit a `BasePattern::Product` in this loop. I'd also make `contains` non-public or rename it to reflect the richer return value.

Fine to do in a followup PR since it's not really related to this one.